### ISSUE TITLE
ClassCastException with failover appender

### ISSUE THREAD

**Issue by cuneytcalishkan** [rfoltyns/log4j2-elasticsearch#40]
**Description**
When using async batch delivery with a failover appender, ClassCastException is thrown when the connection to Elastic server cannot be established.
It seems to be an issue introduced in version 1.4.0, it doesn't exist with 1.3.6

**Configuration**

```xml 
<Elasticsearch name="ElasticsearchAsyncBatch">
    <RollingIndexName indexName="myindex" pattern="yyyy-ww" timeZone="UTC"/>
    <ThresholdFilter level="INFO" onMatch="ACCEPT"/>
    <JacksonJsonLayout>
        <PooledItemSourceFactory poolName="itemPool" itemSizeInBytes="512" initialPoolSize="10000"
                                    monitored="true" monitorTaskInterval="10000" resizeTimeout="500">
            <UnlimitedResizePolicy resizeFactor="0.6"/>
        </PooledItemSourceFactory>
    </JacksonJsonLayout>
    <AsyncBatchDelivery batchSize="10000"
                        deliveryInterval="3000">
        <IndexTemplate name="expense_ng" path="classpath:elasticLogIndexTemplate.json"/>
        <JestBufferedHttp serverUris="http://localhost:9200"
                            connTimeout="500"
                            readTimeout="30000"
                            maxTotalConnection="40"
                            defaultMaxTotalConnectionPerRoute="8"
                            mappingType="index">
            <PooledItemSourceFactory poolName="batchPool" itemSizeInBytes="5120000" initialPoolSize="3"
                                        monitored="true" monitorTaskInterval="10000" resizeTimeout="500">
                <UnlimitedResizePolicy resizeFactor="0.70"/>
            </PooledItemSourceFactory>
        </JestBufferedHttp>
        <AppenderRefFailoverPolicy>
            <AppenderRef ref="Console"/>
        </AppenderRefFailoverPolicy>
    </AsyncBatchDelivery>
</Elasticsearch>
```
**Runtime (please complete the following information):**
 -  log4j2-elasticsearch-jest:1.4.0
 - spring-boot 2.2.5
 - JVM openjdk version "1.8.0_242
 - OS: macOS Mojave 10.14.6
 
**Additional context**
Stackstrace of the exception thrown:
```
2020-03-17 09:57:10,328 restartedMain ERROR IndexTemplate not added: Could not connect to http://localhost:9200
2020-03-17 09:57:10,360 pool-2-thread-1 ERROR Unable to execute failover java.lang.ClassCastException: io.netty.buffer.CompositeByteBuf cannot be cast to java.lang.String
	at org.appenders.log4j2.elasticsearch.AppenderRefFailoverPolicy.deliver(AppenderRefFailoverPolicy.java:41)
	at org.appenders.log4j2.elasticsearch.FailoverPolicy.deliver(FailoverPolicy.java:49)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.concurrent.ConcurrentLinkedQueue$CLQSpliterator.forEachRemaining(ConcurrentLinkedQueue.java:857)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:485)
	at org.appenders.log4j2.elasticsearch.jest.BufferedJestHttpObjectFactory.lambda$createFailureHandler$1(BufferedJestHttpObjectFactory.java:122)
	at org.appenders.log4j2.elasticsearch.jest.BufferedJestHttpObjectFactory$1.failed(BufferedJestHttpObjectFactory.java:157)
	at org.appenders.log4j2.elasticsearch.jest.BufferedJestHttpClient$BufferedResultCallback.failed(BufferedJestHttpClient.java:116)
	at org.apache.http.concurrent.BasicFuture.failed(BasicFuture.java:137)
	at org.apache.http.impl.nio.client.DefaultClientExchangeHandlerImpl.executionFailed(DefaultClientExchangeHandlerImpl.java:101)
	at org.apache.http.impl.nio.client.AbstractClientExchangeHandler.failed(AbstractClientExchangeHandler.java:426)
	at org.apache.http.impl.nio.client.AbstractClientExchangeHandler.connectionRequestFailed(AbstractClientExchangeHandler.java:348)
	at org.apache.http.impl.nio.client.AbstractClientExchangeHandler.access$100(AbstractClientExchangeHandler.java:62)
	at org.apache.http.impl.nio.client.AbstractClientExchangeHandler$1.failed(AbstractClientExchangeHandler.java:392)
	at org.apache.http.concurrent.BasicFuture.failed(BasicFuture.java:137)
	at org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager$1.failed(PoolingNHttpClientConnectionManager.java:316)
	at org.apache.http.concurrent.BasicFuture.failed(BasicFuture.java:137)
	at org.apache.http.nio.pool.RouteSpecificPool.failed(RouteSpecificPool.java:162)
	at org.apache.http.nio.pool.AbstractNIOConnPool.requestFailed(AbstractNIOConnPool.java:613)
	at org.apache.http.nio.pool.AbstractNIOConnPool$InternalSessionRequestCallback.failed(AbstractNIOConnPool.java:893)
	at org.apache.http.impl.nio.reactor.SessionRequestImpl.failed(SessionRequestImpl.java:177)
	at org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor.processEvent(DefaultConnectingIOReactor.java:176)
	at org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor.processEvents(DefaultConnectingIOReactor.java:148)
	at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor.execute(AbstractMultiworkerIOReactor.java:351)
	at org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager.execute(PoolingNHttpClientConnectionManager.java:221)
	at org.apache.http.impl.nio.client.CloseableHttpAsyncClientBase$1.run(CloseableHttpAsyncClientBase.java:64)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by rfoltyns** [comment]
Than you for reporting this. This is due to the changes in failure handling and introduction of `FailedItemSource`.

I've documented this limitation [here](https://github.com/rfoltyns/log4j2-elasticsearch/tree/master/log4j2-elasticsearch-core#appenderreffailoverpolicy).

The simplest fix would be to:
1. implement `toString()` in `StringItemSource`, `ByteBufItemSource` and `FailedItemSource` (last one would have to simply delegate to retain the previous behaviour for now)
2. implement `AppenderRefFailoverPolicy.deliver(ItemSource)`
 
I'll fix in next release.

**Comment by rfoltyns** [comment]
I've implemented a simple patch. I'm planning on releasing it tomorrow.

Would you like to verify that it works as expected?

**Comment by cuneytcalishkan** [comment]
No need to verify, I trust your logic here :)

**Comment by rfoltyns** [comment]
Thanks :) Fixed in 1.4.1.
